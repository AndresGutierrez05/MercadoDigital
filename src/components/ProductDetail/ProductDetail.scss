@use "../../mixins.scss";
@use "sass:map";
@use "../../commonstyles.scss" as common;

.product-detail{
    width: 40%;

    @media screen and (max-width: common.$medium) {
        width: 100%;
    }

    @include mixins.commonProductsContainer;

    .product-selected-img{
        width: 100%;
        display: flex;
        justify-content: center;

        .container-image{
            width: 350px;
            height: 350px;

            @media screen and (max-width: common.$medium) {
                width: 200px;
                height: 200px;
            }
        }

        img{
            width: 100%;
        }
    }

    .product-selected-general{
        display: flex;
        justify-content: space-between;
        width: 100%;

        label{
            color: map.get(common.$theme-colors, "purple");
        }

        .product-selected-add-subtract{
            height: 100%;
            align-self: center;
        }

        img{
            width: 32px;
            padding: 5px;
            cursor: pointer;

            &:hover{
                background: map.get(common.$theme-colors, "orange");
            }
        }
    }
}