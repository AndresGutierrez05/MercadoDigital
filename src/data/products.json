[   
    {
        "Id" : 1,
        "Nombre": "Atún en Lata",
        "Marca" : "Van Camps",
        "Unidad": "g",
        "Descripcion": "Atún en agua",
        "Complemento": "Incluye rallado, lomitos y esparcible",
        "NombrePorcion": "Porción",
        "TamañoPorcion": 80.000,
        "RutaImagen": "https://copservir.vtexassets.com/arquivos/ids/773477-800-auto?v=637951138022000000",
        "Precio" : 5400.0000
    },
    {
        "Id" : 2,
        "Nombre": "Leche Pasteurizada Entera",
        "Marca" : "Alqueria",
        "Unidad": "ml",
        "Descripcion": "Leche de vaca. Alimento básico y equilibrado con excelente densidad nutricional. Aporta proteínas de alto valor biológico, calcio, fósforo, vitaminas A, D y del complejo B.",
        "Complemento": "Cualquier tipo de leche líquida de vaca",
        "NombrePorcion": "Litro",
        "TamañoPorcion": 1300.000,
        "RutaImagen": "https://jumbocolombiaio.vtexassets.com/arquivos/ids/403045-1600-1600",
        "Precio" : 5900.0000
    },
    {
        "Id" : 3,
        "Nombre": "Azúcar Blanca",
        "Marca" : "Riopaila",
        "Unidad": "g",
        "Descripcion": "Azucar blanca refinada",
        "Complemento": "",
        "NombrePorcion": "Bolsa",
        "TamañoPorcion": 1000.000,
        "RutaImagen": "https://vixark.b-cdn.net/lp-i-i-g/azúcar-blanca-riopaila-1000g.jpg",
        "Precio" : 2700.0000
    },
    {
        "Id" : 4,
        "Nombre": "Aceite Vegetal",
        "Marca" : "Frescampo",
        "Unidad": "ml",
        "Descripcion": "Aceite Vegetal FRESCAMPO 3000 ml",
        "Complemento": "",
        "NombrePorcion": "Litro",
        "TamañoPorcion": 3000.000,
        "RutaImagen": "https://exitocol.vtexassets.com/arquivos/ids/13929565-500-auto?v=1753578000",
        "Precio" : 24980.0000
    },
    {
        "Id" : 5,
        "Nombre": "Jabon Avena Barra",
        "Marca" : "Protex",
        "Unidad": "g",
        "Descripcion": "Jabon Avena Barra X3 Unds PROTEX 330 gr",
        "Complemento": "",
        "NombrePorcion": "Barra",
        "TamañoPorcion": 330.000,
        "RutaImagen": "https://exitocol.vtexassets.com/arquivos/ids/14758134-1200-auto?v=637992541550430000&",
        "Precio" : 10900.0000
    },
    {
        "Id" : 6,
        "Nombre": "Atún en Lata",
        "Marca" : "Van Camps",
        "Unidad": "g",
        "Descripcion": "Atún en agua",
        "Complemento": "Incluye rallado, lomitos y esparcible",
        "NombrePorcion": "Porción",
        "TamañoPorcion": 80.000,
        "RutaImagen": "https://copservir.vtexassets.com/arquivos/ids/773477-800-auto?v=637951138022000000",
        "Precio" : 5400.0000
    },
    {
        "Id" : 7,
        "Nombre": "Leche Pasteurizada Entera",
        "Marca" : "Alqueria",
        "Unidad": "ml",
        "Descripcion": "Leche de vaca. Alimento básico y equilibrado con excelente densidad nutricional. Aporta proteínas de alto valor biológico, calcio, fósforo, vitaminas A, D y del complejo B.",
        "Complemento": "Cualquier tipo de leche líquida de vaca",
        "NombrePorcion": "Litro",
        "TamañoPorcion": 1000.000,
        "RutaImagen": "https://jumbocolombiaio.vtexassets.com/arquivos/ids/403045-1600-1600",
        "Precio" : 1900.0000
    },
    {
        "Id" : 8,
        "Nombre": "Azúcar Blanca",
        "Marca" : "Riopaila",
        "Unidad": "g",
        "Descripcion": "Azucar blanca refinada",
        "Complemento": "",
        "NombrePorcion": "Bolsa",
        "TamañoPorcion": 1000.000,
        "RutaImagen": "https://vixark.b-cdn.net/lp-i-i-g/azúcar-blanca-riopaila-1000g.jpg",
        "Precio" : 2700.0000
    },
    {
        "Id" : 9,
        "Nombre": "Aceite Vegetal",
        "Marca" : "Frescampo",
        "Unidad": "ml",
        "Descripcion": "Aceite Vegetal FRESCAMPO 3000 ml",
        "Complemento": "",
        "NombrePorcion": "Litro",
        "TamañoPorcion": 3000.000,
        "RutaImagen": "https://exitocol.vtexassets.com/arquivos/ids/13929565-500-auto?v=1753578000",
        "Precio" : 24980.0000
    },
    {
        "Id" : 10,
        "Nombre": "Jabon Avena Barra",
        "Marca" : "Protex",
        "Unidad": "g",
        "Descripcion": "Jabon Avena Barra X3 Unds PROTEX 330 gr",
        "Complemento": "",
        "NombrePorcion": "Barra",
        "TamañoPorcion": 330.000,
        "RutaImagen": "https://exitocol.vtexassets.com/arquivos/ids/14758134-1200-auto?v=637992541550430000&",
        "Precio" : 10900.0000
    },
    {
        "Id" : 11,
        "Nombre": "Atún en Lata",
        "Marca" : "Van Camps",
        "Unidad": "g",
        "Descripcion": "Atún en agua",
        "Complemento": "Incluye rallado, lomitos y esparcible",
        "NombrePorcion": "Porción",
        "TamañoPorcion": 80.000,
        "RutaImagen": "https://copservir.vtexassets.com/arquivos/ids/773477-800-auto?v=637951138022000000",
        "Precio" : 5400.0000
    },
    {
        "Id" : 12,
        "Nombre": "Leche Pasteurizada Entera",
        "Marca" : "Alqueria",
        "Unidad": "ml",
        "Descripcion": "Leche de vaca. Alimento básico y equilibrado con excelente densidad nutricional. Aporta proteínas de alto valor biológico, calcio, fósforo, vitaminas A, D y del complejo B.",
        "Complemento": "Cualquier tipo de leche líquida de vaca",
        "NombrePorcion": "Litro",
        "TamañoPorcion": 1000.000,
        "RutaImagen": "https://jumbocolombiaio.vtexassets.com/arquivos/ids/403045-1600-1600",
        "Precio" : 1900.0000
    },
    {
        "Id" : 13,
        "Nombre": "Azúcar Blanca",
        "Marca" : "Riopaila",
        "Unidad": "g",
        "Descripcion": "Azucar blanca refinada",
        "Complemento": "",
        "NombrePorcion": "Bolsa",
        "TamañoPorcion": 1000.000,
        "RutaImagen": "https://vixark.b-cdn.net/lp-i-i-g/azúcar-blanca-riopaila-1000g.jpg",
        "Precio" : 2700.0000
    },
    {
        "Id" : 14,
        "Nombre": "Aceite Vegetal",
        "Marca" : "Frescampo",
        "Unidad": "ml",
        "Descripcion": "Aceite Vegetal FRESCAMPO 3000 ml",
        "Complemento": "",
        "NombrePorcion": "Litro",
        "TamañoPorcion": 3000.000,
        "RutaImagen": "https://exitocol.vtexassets.com/arquivos/ids/13929565-500-auto?v=1753578000",
        "Precio" : 24980.0000
    },
    {
        "Id" : 15,
        "Nombre": "Jabon Avena Barra",
        "Marca" : "Protex",
        "Unidad": "g",
        "Descripcion": "Jabon Avena Barra X3 Unds PROTEX 330 gr",
        "Complemento": "",
        "NombrePorcion": "Barra",
        "TamañoPorcion": 330.000,
        "RutaImagen": "https://exitocol.vtexassets.com/arquivos/ids/14758134-1200-auto?v=637992541550430000&",
        "Precio" : 10900.0000
    }
]