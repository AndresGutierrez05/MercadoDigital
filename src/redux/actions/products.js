export const START_GET_PRODUCTS = 'START_GET_PRODUCTS';
export const SUCCESS_GET_PRODUCTS = 'SUCCESS_GET_PRODUCTS';
export const START_GET_PRODUCT = 'START_GET_PRODUCT';
export const SUCCESS_GET_PRODUCT = 'SUCCESS_GET_PRODUCT';
export const START_REMOVE_SELECTED_PRODUCT = 'START_REMOVE_SELECTED_PRODUCT';
export const SUCCESS_REMOVE_SELECTED_PRODUCT = 'SUCCESS_REMOVE_SELECTED_PRODUCT';
export const START_SET_CART = 'START_SET_CART';
export const SUCCESS_SET_CART = 'SUCCESS_SET_CART';
export const START_REMOVE_PRODUCT_CART = 'START_REMOVE_PRODUCT_CART';
export const SUCCESS_REMOVE_PRODUCT_CART = 'SUCCESS_REMOVE_PRODUCT_CART';
export const START_ADD_PRODUCT_QUANTITY = 'START_ADD_PRODUCT_QUANTITY';
export const SUCCESS_ADD_PRODUCT_QUANTITY = 'SUCCESS_ADD_PRODUCT_QUANTITY';
export const START_DISCOUNT_PRODUCT_QUANTITY = 'START_DISCOUNT_PRODUCT_QUANTITY';
export const SUCCESS_DISCOUNT_PRODUCT_QUANTITY = 'SUCCESS_DISCOUNT_PRODUCT_QUANTITY';

export const startGetProducts = payload => ({
    type: START_GET_PRODUCTS,
    ...payload
});

export const startProductSelected = payload => ({
    type: START_GET_PRODUCT,
    ...payload
});

export const startRemoveProductCart = payload => ({
    type: START_REMOVE_PRODUCT_CART,
    ...payload
});

export const startSetCart = payload => ({
    type: START_SET_CART,
    ...payload
});

export const startRemoveSelectedProduct = payload => ({
    type: START_REMOVE_SELECTED_PRODUCT,
    ...payload
});

export const startAddSelectedProduct = payload => ({
    type: START_ADD_PRODUCT_QUANTITY,
    ...payload
});

export const startDiscountSelectedProduct = payload => ({
    type: START_DISCOUNT_PRODUCT_QUANTITY,
    ...payload
});