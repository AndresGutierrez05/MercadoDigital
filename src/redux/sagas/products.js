import { put, takeLatest } from 'redux-saga/effects';
import { START_GET_PRODUCTS, SUCCESS_GET_PRODUCTS, START_GET_PRODUCT, START_REMOVE_PRODUCT_CART, START_REMOVE_SELECTED_PRODUCT, START_ADD_PRODUCT_QUANTITY, START_DISCOUNT_PRODUCT_QUANTITY, START_SET_CART, 
   SUCCESS_GET_PRODUCT, SUCCESS_SET_CART, SUCCESS_REMOVE_PRODUCT_CART, SUCCESS_REMOVE_SELECTED_PRODUCT, SUCCESS_ADD_PRODUCT_QUANTITY, SUCCESS_DISCOUNT_PRODUCT_QUANTITY  } from '../actions/products';
import productsJSON from "../../data/products.json";

function* getProducts() {
   try {
      yield put({type: SUCCESS_GET_PRODUCTS, payload: productsJSON});
   } catch (e) {
      yield put({type: SUCCESS_GET_PRODUCTS, message: e.message});
   }
}

function* setProductSelected(action){
   try {
      yield put({type: SUCCESS_GET_PRODUCT, payload: action});
   } catch (e) {
      yield put({type: SUCCESS_GET_PRODUCT, message: e.message});
   }
}

function* setCart(action) {
   try {
      yield put({type: SUCCESS_SET_CART, payload: action});
   } catch (e) {
      yield put({type: SUCCESS_SET_CART, message: e.message});
   }
}

function* removeProductCart(action) {
   try {
      yield put({type: SUCCESS_REMOVE_PRODUCT_CART, payload: action});
   } catch (e) {
      yield put({type: SUCCESS_REMOVE_PRODUCT_CART, message: e.message});
   }
}

function* removeSelectedProduct(action) {
   try {
      yield put({type: SUCCESS_REMOVE_SELECTED_PRODUCT, payload: action});
   } catch (e) {
      yield put({type: SUCCESS_REMOVE_SELECTED_PRODUCT, message: e.message});
   }
}

function* addProductQuantity(action) {
   try {
      yield put({type: SUCCESS_ADD_PRODUCT_QUANTITY, payload: action});
   } catch (e) {
      yield put({type: SUCCESS_ADD_PRODUCT_QUANTITY, message: e.message});
   }
}

function* discountProductQuantity(action) {
   try {
      yield put({type: SUCCESS_DISCOUNT_PRODUCT_QUANTITY, payload: action});
   } catch (e) {
      yield put({type: SUCCESS_DISCOUNT_PRODUCT_QUANTITY, message: e.message});
   }
}


// Watchers
export default function* products() {
  yield takeLatest(START_GET_PRODUCTS, getProducts);
  yield takeLatest(START_GET_PRODUCT, setProductSelected);
  yield takeLatest(START_SET_CART, setCart);
  yield takeLatest(START_REMOVE_PRODUCT_CART, removeProductCart);
  yield takeLatest(START_REMOVE_SELECTED_PRODUCT, removeSelectedProduct);
  yield takeLatest(START_ADD_PRODUCT_QUANTITY, addProductQuantity);
  yield takeLatest(START_DISCOUNT_PRODUCT_QUANTITY, discountProductQuantity);
}