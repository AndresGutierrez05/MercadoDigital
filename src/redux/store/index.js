import { createStore, applyMiddleware, combineReducers } from 'redux';
import reduxSaga from 'redux-saga';
import rootSaga from '../sagas';
import productsReducer from "../reducers/productsReducer.js";

const rootReducer = combineReducers({
    productosReducer: productsReducer,
});

const sagaMiddleware = reduxSaga();

export default () => {
    return {
        ...createStore(rootReducer, applyMiddleware(sagaMiddleware)),
        runSaga: sagaMiddleware.run(rootSaga)
    }
};