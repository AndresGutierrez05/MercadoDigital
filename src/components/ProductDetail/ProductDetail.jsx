import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from 'react-redux';
import { startAddSelectedProduct, startDiscountSelectedProduct } from "../../redux/actions/products";
import ImageProduct from "../ImageProduct/ImageProduct";
import "./ProductDetail.scss"

export default function ProductDetail({ propWindowSize, propProduct }){
    const dispatch = useDispatch();
    const reduxProduct = useSelector((state) => state.productosReducer.productoSeleccionado);
    const product = propProduct || reduxProduct;
    const cart = useSelector((state) => state.productosReducer.carrito) ?? [];
    
    const getWindowSize = _ => {
        const {innerWidth} = window;
        return innerWidth;
    }
    
    const [windowSize, setWindowSize] = useState(propWindowSize || getWindowSize())
    const addQuantity = id => {
        dispatch(startAddSelectedProduct({id : id}));
    };

    const discountQuantity = id => {
        dispatch(startDiscountSelectedProduct({id : id}));
    };

    useEffect(() => {
        function handleWindowResize() {
            setWindowSize(getWindowSize());
        }
    
        window.addEventListener('resize', handleWindowResize);
    
        return () => {
          window.removeEventListener('resize', handleWindowResize);
        };
      }, []);

    return(
            <div className="product-detail">
                <h2>Detalle del producto</h2>
                <hr/>
                {product != null 
                ? 
                    <div className="product-selected">
                        <div className="product-selected-img">
                            <ImageProduct src={product.RutaImagen} alt={product.Nombre} quantity={cart.filter(productSelected => productSelected.Id === product.Id).length > 0 ? cart.filter(productSelected => productSelected.Id === product.Id)[0].cantidad : 0 }/>
                        </div>
                        <br></br>
                        <div className="product-selected-general">
                            <div className="product-selected-info">
                                <h3>{product.Nombre}&nbsp;&nbsp;<label>${product.Precio}</label></h3>
                            </div>
                            <div className="product-selected-add-subtract">
                                <img src="/icons/minus.png" onClick={() => discountQuantity(product.Id)} alt="Restar"></img>
                                <img src="/icons/plus.png" onClick={() => addQuantity(product.Id)} alt="Sumar"></img>
                            </div>
                        </div>
                        <hr/>
                            {product.Descripcion}
                        <hr/>
                    </div>
                :
                    <p>Por favor seleccione un producto de la parte { windowSize <= 900 ? "de arriba" : "izquierda" } para ver su detalle y asignar mas</p>
                }
            </div>
    )
}