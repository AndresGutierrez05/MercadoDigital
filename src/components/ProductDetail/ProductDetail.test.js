import ProductDetail from "./ProductDetail.jsx";
import renderer from "react-test-renderer";
import store from "../../redux/store/index.js";
import { Provider } from "react-redux";
import products from "../../data/products.json";

describe('Product detail component', () => {
    test("default component test (phone-size)", () => {
        const component = renderer.create(
            <Provider store={store()}>
                <ProductDetail propWindowSize={400}/>,
            </Provider>
            );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    test("default component test (nomal-size)", () => {
        const component = renderer.create(
            <Provider store={store()}>
                <ProductDetail propWindowSize={1000} />,
            </Provider>
            );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    }); 

    if(products !== null && products.length > 0){
        test("Product detail first product component test", () => {
            const component = renderer.create(
                <Provider store={store()}>
                    <ProductDetail propWindowSize={1000} propProduct={products[0]}/>,
                </Provider>
                );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        }); 
    }
});