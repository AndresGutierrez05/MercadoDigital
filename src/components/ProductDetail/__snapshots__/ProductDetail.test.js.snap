// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product detail component Product detail first product component test 1`] = `
Array [
  <div
    className="product-detail"
  >
    <h2>
      Detalle del producto
    </h2>
    <hr />
    <div
      className="product-selected"
    >
      <div
        className="product-selected-img"
      >
        <div
          className="container-image"
        >
          <div
            className="image"
          >
            <img
              alt="Atún en Lata"
              src="https://copservir.vtexassets.com/arquivos/ids/773477-800-auto?v=637951138022000000"
            />
          </div>
        </div>
      </div>
      <br />
      <div
        className="product-selected-general"
      >
        <div
          className="product-selected-info"
        >
          <h3>
            Atún en Lata
              
            <label>
              $
              5400
            </label>
          </h3>
        </div>
        <div
          className="product-selected-add-subtract"
        >
          <img
            alt="Restar"
            onClick={[Function]}
            src="/icons/minus.png"
          />
          <img
            alt="Sumar"
            onClick={[Function]}
            src="/icons/plus.png"
          />
        </div>
      </div>
      <hr />
      Atún en agua
      <hr />
    </div>
  </div>,
  ",",
]
`;

exports[`Product detail component default component test (nomal-size) 1`] = `
Array [
  <div
    className="product-detail"
  >
    <h2>
      Detalle del producto
    </h2>
    <hr />
    <p>
      Por favor seleccione un producto de la parte 
      izquierda
       para ver su detalle y asignar mas
    </p>
  </div>,
  ",",
]
`;

exports[`Product detail component default component test (phone-size) 1`] = `
Array [
  <div
    className="product-detail"
  >
    <h2>
      Detalle del producto
    </h2>
    <hr />
    <p>
      Por favor seleccione un producto de la parte 
      de arriba
       para ver su detalle y asignar mas
    </p>
  </div>,
  ",",
]
`;
